### PR TITLE
AI - Much better burning/daylight behavior

### PR DESCRIPTION
**Pull request by SafwatHalaby**
Mobs will stand their ground if it means they won't burn there and they'll burn if they chase you.

Mobs WONT stand their ground if you are in a shade and they are in a shade and they can walk to you.

Mobs WONT stand their ground if  they are going to burn to death anyways.

Mobs will attack regardless of the light condition if they were provoked recently.

No more zombie freezing at twilight.

When very close to them, they sometimes step out of the shade accidentally, this was something I was planning to fix but it's actually a neat feature :wink:

### CODE REVIEW

File: src/Mobs/Monster.cpp [hunk 1]
```diff
@@ -300,7 +300,12 @@ void cMonster::Tick(std::chrono::milliseconds a_Dt, cChunk & a_Chunk)
 		{
 			if (--m_GiveUpCounter == 0)
 			{
-				FinishPathFinding();
+				FinishPathFinding();  // Not to be confused with StopMovingToPosition, this just discards the current path and calculates another.
```
Comment (madmaxoft) [inline src/Mobs/Monster.cpp]: How about calling it `RestartPathFinding`, then? Or something along these lines.

File: src/Mobs/Monster.cpp [hunk 2]
```diff
@@ -1129,7 +1134,7 @@ bool cMonster::WouldBurnAt(Vector3d a_Location, cChunk & a_Chunk)
 		(a_Chunk.GetSkyLight(RelX, RelY, RelZ) == 15) &&             // In the daylight
 		(a_Chunk.GetBlock(RelX, RelY, RelZ) != E_BLOCK_SOULSAND) &&  // Not on soulsand
 		(GetWorld()->GetTimeOfDay() < (12000 + 1000)) &&             // It is nighttime
-		!IsOnFire() &&                                               // Not already burning
+
```
Comment (madmaxoft) [inline src/Mobs/Monster.cpp]: Empty line inside a condition? That doesn't look too good.

File: src/Mobs/Monster.cpp (hunk 1 above)
Comment (wiseoldman95) [inline src/Mobs/Monster.cpp]: It's not always used for restarting though, I guess `ResetPathFinding` is best.

Comment (wiseoldman95) [inline src/Mobs/Monster.cpp]: Got a name which means "Either finish or restart, depending on whether we're done or not"?

File: src/Mobs/Monster.cpp (hunk 2 above)
Comment (wiseoldman95) [inline src/Mobs/Monster.cpp]: Done and rebased.

**Comment by wiseoldman95** [comment]
Standing by for comments / Rebase and merge.

File: src/Mobs/Monster.cpp [hunk 3]
```diff
@@ -1129,7 +1135,6 @@ bool cMonster::WouldBurnAt(Vector3d a_Location, cChunk & a_Chunk)
 		(a_Chunk.GetSkyLight(RelX, RelY, RelZ) == 15) &&             // In the daylight
```
Comment (tigerw) [inline src/Mobs/Monster.cpp]: Unrelated, but the usage of chunks are causing issues, since I believe mobs can move to a different chunk (invalidating the parameter) in super::Tick, and then this function is called.

Fix is to use the GET_AND_VERIFY_CURRENT_CHUNK macro and place it under super::Tick in this file, I think.

Comment (wiseoldman95) [inline src/Mobs/Monster.cpp]: Super already has this line: `GET_AND_VERIFY_CURRENT_CHUNK(NextChunk, POSX_TOINT, POSZ_TOINT)`, wouldn't that solve it?

Comment (wiseoldman95) [inline src/Mobs/Monster.cpp]: I fixed it in this particular case, but this could happen ANYWHERE, lots of stuff, including many mob types, inhert the tick() function.

Comment (tigerw) [inline src/Mobs/Monster.cpp]: Yeah. We have many bugs :P

**Comment by wiseoldman95** [comment]
Squash?

File: src/Mobs/Monster.cpp (hunk 3 above)
Comment (wiseoldman95) [inline src/Mobs/Monster.cpp]: We could prevent this with a macro, instead of calling super::tick, make two macros such as:
`SUPER_TICK_AT_START();`
and 
`SUPER_TICK_END()`
The former first does the chunk update thing, and only then calls super, while the latter just calls super.

Comment (wiseoldman95) [inline src/Mobs/Monster.cpp]: The first one is used when you want to call SUPER at the begining of your tick, the latter when you want to call it at the end of your tick.

Comment (wiseoldman95) [inline src/Mobs/Monster.cpp]: Or have a single one called Super_safeTick or something

**Comment by wiseoldman95** [comment]
Squashed.

File: src/Mobs/Monster.cpp (hunk 3 above)
Comment (tigerw) [inline src/Mobs/Monster.cpp]: Macros everywhere! The main thing would be to remember to highlight to contributors that it needs to be done if they are using the chunk parameter. I suppose cMonster::Tick could call GET_AND_VERIFY... before it does work and after to account for all situations, but for some mobs the extra processing is needless since they don't override.

Comment (wiseoldman95) [inline src/Mobs/Monster.cpp]: Alright, but that'll be in another PR. This one is ready for merging.

**Comment by SafwatHalaby** [comment]
To avoid confusion: My username has been changed from WiseOldMan95 to SafwatHalaby.
